Lookup for new well-known types to fail if duplicates are found

jcouv opened the pull request:
The lookup for well-known types currently succeeds (with a `WRN_MultiplePredefTypes` warning) when more than one candidate is found.
Per LDM discussion last week, well-known types introduced with C# 7 or later will adopt a stricter behavior. Their lookup will fail if the result is ambiguous.

Overview: 
This change introduces `CSharp7Sentinel` as a marker in the `WellKnownType` enum. The `GetWellKnownType` method passes a new `allowAmbiguity` flag during the lookup. 
The method that used to issue a warning, now could generate an error.
Also, to fix #11322, we'll lookup the proper VT types even for tuple literals that don't have a natural type.

I didn't have to add a test to verify the old behavior (for old well-known types), as we already have a couple.

Fixes #11327 and #11322

CC @dotnet/roslyn-compiler for review.

cston commented on src/Compilers/CSharp/Portable/Errors/ErrorCode.cs:
```diff
@@ -1391,5 +1391,6 @@ internal enum ErrorCode
         ERR_TupleExplicitNamesOnAllMembersOrNone = 8204,
 
         ERR_PredefinedTypeMemberNotFoundInAssembly = 8205,
+        ERR_PredefinedTypeNotUniquelyFound = 8206,
```
Perhaps `ERR_PredefinedTypeAmbiguous` or `ERR_AmbiguousPredefinedType`.

jcouv replied:
@cston Hold on on further review. From discussion with Aleksey just now, I am making a significant simplification. We'll return null instead and not introduce a new error.

jcouv commented:
@cston Just pushed the simplification I mentioned. Re-using existing error message is ok, although a little bit less informative.

cston commented on src/Compilers/CSharp/Portable/Symbols/Compilation_WellKnownMembers.cs:
```diff
@@ -97,8 +97,18 @@ internal NamedTypeSymbol GetWellKnownType(WellKnownType type)
                 {
-                    result = this.Assembly.GetTypeByMetadataName(
-                        mdName, includeReferences: true, useCLSCompliantNameArityEncoding: true, isWellKnownType: true, warnings: warnings);
+                    if (type <= WellKnownType.CSharp7Sentinel)
```
Consider using conditional on `warnings` parameter in call: `warnings: (type < Sentinel) ? null : warnings`.

jcouv replied:
I left this one as-is because I found it hard to read.

cston replied:
Duplicating the line means the reader has to spot the differences between the two and also increases the chance the two cases will get out of sync.

jcouv replied:
That's a good point. Would introducing a local be ok?

cston replied:
Sounds good, thanks.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Tuples/TupleTypeSymbol.cs:
```diff
@@ -369,6 +369,33 @@ private static NamedTypeSymbol GetTupleUnderlyingType(ImmutableArray<TypeSymbol>
+
+            NamedTypeSymbol firstTupleType = compilation.GetWellKnownType(GetTupleType(remainder));
+
+            if ((object)diagnostics != null && (object)syntax != null)
```
It feels like the whole purpose of this method is to report errors when types are missing. Consider asserting that diagnostics is not null on entry and dropping the null checks.

AlekseyTs replied:
Same for syntax.

---
In reply to: [63761762](https://github.com/dotnet/roslyn/pull/11389#discussion_r63761762) [](ancestors = 63761762)

jcouv replied:
Done. Thanks

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Tuples/TupleTypeSymbol.cs:
```diff
@@ -369,6 +369,33 @@ private static NamedTypeSymbol GetTupleUnderlyingType(ImmutableArray<TypeSymbol>
+                {
+                    Binder.ReportUseSiteDiagnostics(chainedTupleType, diagnostics, syntax);
+
+                }
```
An unnecessary empty line above.

jcouv replied:
Removed

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/MissingSpecialMember.cs:
```diff
@@ -572,6 +572,233 @@ public void AllWellKnownTypes()
+
+            Assert.True(WellKnownType.System_Environment <= WellKnownType.CSharp7Sentinel);
+
+            Assert.True(WellKnownType.System_Runtime_GCLatencyMode <= WellKnownType.CSharp7Sentinel);
```
Are all members in WellKnownType tightly packed? Consider asserting actual difference between First and CSharp7Sentinel in order to make sure that none of the new types are inserted into the range.

cston replied:
```foreach (var type = new [] { WellKnownType.System_Math, ..., WellKnownType.GCLatencyMode }) Assert.True(type ...);```

jcouv replied:
Done and done.

This made me wonder. Should there be a similar change for special types too?

AlekseyTs replied:
I don't think they were ever affected. There could be no ambiguity with special types because they are always looked up in one assembly.

jcouv replied:
Makes sense. Thanks

AlekseyTs commented:
LGTM

VSadov commented:
LGTM

cston commented:
LGTM

jcouv commented:
Thanks for the review.
The latest commit I pushed addressed the suggestions. 

As Aleksey suggested, I copied the list of well-known types from `master` branch (authority on CSharp6) and found that one type had already moved. This is corrected now.

AlekseyTs commented:
LGTM
